kubernetes/resource_kubernetes_tokenrequest.go

```diff
@@ -0,0 +1,85 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package kubernetes
+
+import (
+	"context"
+	"log"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	apiv1 "k8s.io/api/authentication/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func resourceKubernetesTokenRequestV1() *schema.Resource {
+	return &schema.Resource{
+		CreateContext: resourceKubernetesTokenRequestCreateV1,
+		ReadContext:   resourceKubernetesTokenRequestReadV1,
+		UpdateContext: resourceKubernetesTokenRequestUpdateV1,
+		DeleteContext: resourceKubernetesTokenDeleteV1,
+
+		Schema: map[string]*schema.Schema{
+			"metadata": namespacedMetadataSchema("token request", true),
+			"spec": {
+				Type:        schema.TypeList,
+				Description: apiv1.TokenRequest{}.Spec.SwaggerDoc()["spec"],
+				Required:    true,
+				MaxItems:    1,
+				Elem: &schema.Resource{
+					Schema: tokenRequestSpecFields(),
+				},
+			},
+			"token": {
+				Type:        schema.TypeString,
+				Description: "Token is the opaque bearer token.",
+				Computed:    true,
+				Sensitive:   true,
+			},
+		},
+	}
+}
+
+func resourceKubernetesTokenRequestCreateV1(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	conn, err := meta.(KubeClientsets).MainClientset()
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	metadata := expandMetadata(d.Get("metadata").([]interface{}))
+	spec := expandTokenRequestSpec(d.Get("spec").([]interface{}))
+	saName := d.Get("metadata.0.name").(string)
+
+	request := apiv1.TokenRequest{
+		ObjectMeta: metadata,
+		Spec:       *spec,
+	}
+
+	log.Printf("[INFO] Creating new TokenRequest: %#v", request)
+	out, err := conn.CoreV1().ServiceAccounts(metadata.Namespace).CreateToken(ctx, saName, &request, metav1.CreateOptions{})
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	d.Set("token", out.Status.Token)
+
```

Comment:
You also need to take `out.Spec` and run it through your flattenTokenRequestSpec function and set a value for `spec`. You have to do this because if you don't specify values for these then the API server fills them in with default values, e.g on cluster I get the following default values:

```
"audiences": [
  "https://kubernetes.default.svc.cluster.local"
],
"expirationSeconds": 3600,
"boundObjectRef": null
 ```